pipelines/pspec_pipeline/pspec_pipe.py

```diff
@@ -0,0 +1,228 @@
+# Run OQE Pipeline
+#-------------------------------------------------------------------------------
+if run_pspec:
+    # get algorithm parameters
+    globals().update(cf['algorithm']['pspec'])
+    time = datetime.utcnow()
+    hp.utils.log("\n{}\nstarting pspec OQE: {}\n".format("-"*60, time), f=lf, verbose=verbose)
+
+    # configure dataset groupings
+    groupings = hp.utils.config_pspec_blpairs(data_template, pol_pairs, group_pairs, exclude_auto_bls=exclude_auto_bls,
+                                              exclude_permutations=exclude_permutations, bl_len_range=bl_len_range,
+                                              bl_deg_range=bl_deg_range, xants=xants)
+
+    # create dictionary of individual jobs to launch
+    jobs = odict()
+    for pair in groupings:
+        # make Nblps / Nblps_per_job jobs for this pair
+        blps = groupings[pair]
+        Nblps = len(blps)
+        Njobs = Nblps // Nblps_per_job + 1
+        job_blps = [blps[i*Nblps_per_job:(i+1)*Nblps_per_job] for i in range(Njobs)]
+        for i, _blps in enumerate(job_blps):
+            key = tuple(hp.utils.flatten(pair) + [i])
+            jobs[key] = _blps
+
+    # create pspec worker function
+    def pspec(i, outfile=outfname, dt=data_template, st=std_template, jobs=jobs, pol_pairs=pol_pairs, p=cf['algorithm']['pspec']):
+        try:
+            # get key
+            key = jobs.keys()[i]
+            # parse dsets
+            dsets = [dt.format(group=key[0], pol=key[2]), dt.format(group=key[1], pol=key[3])]
+            dset_labels = [os.path.basename(d) for d in dsets]
+            if st is not None and st is not '' and st is not 'None':
+                dsets_std = [dt.format(group=key[0], pol=key[2]), dt.format(group=key[1], pol=key[3])]
+            else:
+                dsets_std = None
+
+            # pspec_run
+            hp.pspecdata.pspec_run(dsets, outfile, dsets_std=dsets_std, dset_labels=dset_labels,
+                                   dset_pairs=[(0, 1)], spw_ranges=p['spw_ranges'], n_dlys=p['n_dlys'],
+                                   pol_pairs=pol_pairs, blpairs=jobs[key], input_data_weight=p['input_data_weight'],
+                                   norm=p['norm'], taper=p['taper'], beam=p['beam'], cosmo=p['cosmo'],
+                                   rephase_to_dset=p['rephase_to_dset'], trim_dset_lsts=p['trim_dset_lsts'],
+                                   broadcast_dset_flags=p['broadcast_dset_flags'], time_thresh=p['time_thresh'],
+                                   Jy2mK=p['Jy2mK'], overwrite=overwrite, psname_ext="_{}".format(key[4]),
+                                   verbose=verbose)
+        except:
+            hp.utils.log("\nPSPEC job {} errored with:".format(i), f=ef, tb=sys.exc_info(), verbose=verbose)
+            return 1
+
+        return 0
+
+    # launch pspec jobs
+    failures = hp.utils.job_monitor(pspec, range(len(jobs)), "PSPEC", lf=lf, maxiter=maxiter, verbose=verbose)
```

Comment:
Does this do any timing? If not, it might be neat to print some timing info so we can check on progress and estimate how long things are likely to run for.

Reply:
Sure I can add some basic timing into it

Reply:
So each block in the pspec pipeline already has a timer, so we can do small runs and get a sense for how long each block takes.